File: LeetCode/Python Solutions/Longest Common Subsequence/longest.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        text1 = tuple(reversed(text1))
        text2 = tuple(reversed(text2))

        m = len(text1)
        n = len(text2)

        memo = [0] * (n + 1)
        for i in range(1, m + 1):
            diag = 0
            for j in range(1, n + 1):
                tmp = memo[j]
                memo[j] = diag + 1 if text1[i - 1] == text2[j - 1] else max(memo[j], memo[j - 1])
                diag = tmp

        return memo[n]

# Tests
if __name__ == '__main__':
    Sol = Solution()
    Solve = Sol.longestCommonSubsequence(text1 = "abc", text2 = "abc" ) 
    # text1 = "abc", text2 = "abc" -> 3
    print(Solve)
    Solve = Sol.longestCommonSubsequence(text1 = "abc", text2 = "def" )  
    # text1 = "abc", text2 = "def" -> 0
    print(Solve)

File: LeetCode/Java Solutions/Longest Common Subsequence/longest.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        // Reverse the strings
        text1 = new StringBuilder(text1).reverse().toString();
        text2 = new StringBuilder(text2).reverse().toString();

        int m = text1.length();
        int n = text2.length();

        // Initialize memo array
        int[] memo = new int[n + 1];

        for (int i = 1; i <= m; i++) {
            int diag = 0;
            for (int j = 1; j <= n; j++) {
                int tmp = memo[j];
                // Update memo[j] based on whether characters match
                memo[j] = (text1.charAt(i - 1) == text2.charAt(j - 1)) ? diag + 1 : Math.max(memo[j], memo[j - 1]);
                diag = tmp;
            }
        }

        return memo[n];
    }
}
